File: LeetCode/0877/golang/lc0877.go
package lc0877

// Runtime complexity: O(1)
// Auxiliary space: O(1)
// Subjective level: medium.
// Solved on: 2026-08-02
func stoneGame(piles []int) bool {
	// Given that the number of piles is even (so important!),
	// Alice can always win, given that she plays optimally.
	// This is because she can in effect: select the *best* pile, on move 1.
	// Bob would have a chance at winning only at "impossible to win" scenarios,
	// such as []int{1, 101, 1}; but these are not allowed as the total piles' count is even!
	// So in effect, Alice wins in every scenario.
	return true
}

File: LeetCode/0877/cpp/lc0877.cpp
#include <iostream>
#include <vector>

// Runtime complexity: O(1)
// Auxiliary space complexity: O(1)
// Subjective level: medium.
// Solved on: 2026-08-02
class Solution {
public:
    bool stoneGame(std::vector<int>& piles) {
        return true;
    }
};

int main() {
    struct Testcase {
        std::vector<int> piles;
        bool expected;
    };
    Testcase testcases[] = {
        {{5, 3, 4 ,5}, true},
        {{3, 7, 2, 3}, true},
    };
    Solution s{};
    int numGood = 0, numBad = 0;
    for (Testcase& tc: testcases) {
        auto actual = s.stoneGame(tc.piles);
        if (actual != tc.expected) {
            std::cout << "Testcase " << (numGood+numBad) << " failed. Got: " << actual
                << ", want: " << tc.expected << "\n";
            ++numBad;
        } else {
            ++numGood;
        }
    }
    std::cout << (numBad == 0 ? "[OK]" : "[FAIL]") << " "
        << numGood << "/" << (numBad + numGood) << " testcases passed successfully.\n";
}
